### PR TITLE
FT: Move replay lock to Redis

### PR DESCRIPTION
**Pull request by bennettbuchanan**

### CODE REVIEW

**Comment by bennettbuchanan** [comment]
@ironman-machine

File: src/lib/Datastore.js [hunk 1]
```diff
@@ -37,6 +37,26 @@ export default class Datastore {
     }
 
     /**
+    * Set the replay lock key (if it does not already exist) with an expiration
+    * of fifteen minutes
+    * @return {undefined}
+    */
+    setLock() {
+        const ttl = 60 * 15; // fifteen minutes
```
Comment (rahulreddy) [inline src/lib/Datastore.js]: Datastore method should be generic. You can call this setExpire which takes a key, value and ttl as params.

File: src/lib/Datastore.js [hunk 2]
```diff
@@ -37,6 +37,26 @@ export default class Datastore {
...
+        // This method is a Promise because no callback is given.
+        return this._client.set('s3:utapireplay:lock', 'true', 'EX', ttl, 'NX');
+    }
+
+    /**
+    * Set the replay lock key if it exists. If there is an error during this
+    * command, do not handle it as the lock will expire after fifteen minutes.
+    * @return {undefined}
+    */
+    removeLock() {
```
Comment (rahulreddy) [inline src/lib/Datastore.js]: Same here. `delete` should be used to keep it generic.

File: src/lib/UtapiReplay.js [hunk 3]
```diff
@@ -187,11 +186,14 @@ export default class UtapiReplay {
         }
         const replay = scheduleJob(this.replaySchedule, () => {
             // Ensure no elements are still being pushed by a prior replay.
-            if (!this.replayLock) {
-                this._setReplayLock(true);
-                return this._checkLocalCache();
-            }
-            return undefined;
+            this.localCache.setExpire('s3:utapireplay:lock', 'true', TTL)
```
Comment (rahulreddy) [inline src/lib/UtapiReplay.js]: I liked the original approach of using an abstract method for locking `_setReplayLock()`

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by rahulreddy** [comment]
Please squash, I will merge.

**Comment by bennettbuchanan** [comment]
@ironman-machine

**Comment by bennettbuchanan** [comment]
E2E passed.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Pull request merged**